### DISCUSSION TITLE
How do I write unit tests?

### DISCUSSION THREAD

**Discussion by maikelos** [xonsh/xonsh#5303] [Q&A] [answered]
The documentation says nothing about how to write unit tests for .xsh scripts. I reckon pytest can be used for this purpose but there is no information provided on how to set it up such that it doesn't get tripped on xonsh specific syntax.

**Comment by gforsyth** [comment, marked as answer]
Yeah, this should be documented. 

You have to use the `XonshImportHook` and a bit of `meta_path` hacking, so you do:

```xsh
import sys

from xonsh.imphooks import XonshImportHook
from xonsh.execer import Execer

sys.meta_path.append(XonshImportHook(Execer()))
```

at the top of your `test_` file. And the tests can be written in either a `.py` file or a `.xsh` file.

Gist here for full example: https://gist.github.com/gforsyth/b07799771dd530ee51ad6bc3068b25bb

**Reply by anki-code** [reply]
Thanks!

@maikelos feel free to create PR with updating [docs](https://github.com/xonsh/xonsh/tree/main/docs).
